easydock/preparation_for_docking.py

```diff
@@ -35,43 +38,246 @@ def mol_from_smi_or_molblock(ligand_string):
+            ssr += d * d
+        ssr /= mol.GetNumAtoms()
+
+    return np.sqrt(ssr)
+
+def GetConformerRMSMatrixForSaturatedRingMolecule(mol: Chem.Mol, atomIds:list[list[int]]=None, prealigned: bool=False) -> list[float]:
+    """ Returns the RMS matrix of the conformers of a molecule based on the alignment and rmsd of saturated ring.
+        The function calculates the mean of the RMSD of each saturated ring with the GetConformerRMSFromAtomIds.
+        The alignment is done per ring (for example, three alignments are done for three saturated ring) 
+        
+        By default, the conformers will be aligned to the first conformer
+        before the RMS calculation and, as a side-effect, the second will be left
+        in the aligned state.
+
+    As a side-effect, the conformers will be aligned to the first
+    conformer (i.e. the reference) and will left in the aligned state.
+
+    Arguments:
+        - mol:     the molecule
+        - atomIds: (optional) list of atom ids to use a points for
+                    alingment - defaults to all atoms
+        - prealigned: (optional) by default the conformers are assumed
+                        be unaligned and will therefore be aligned to the
+                        first conformer
+
+    Note that the returned RMS matrix is symmetrical, i.e. it is the
+    lower half of the matrix, e.g. for 5 conformers::
+
+        rmsmatrix = [ a,
+                      b, c,
+                      d, e, f,
+                      g, h, i, j]
+
+    where a is the RMS between conformers 0 and 1, b is the RMS between
+    conformers 0 and 2, etc.
+    This way it can be directly used as distance matrix in e.g. Butina
+    clustering.
 
-    def gen_conf(mole, useRandomCoords, randomSeed):
+    """
+
+    cmat_list = []
+    for atom_id_in_ring in atomIds:
+        # if necessary, align the conformers
+        # Note: the reference conformer is always the first one
+        rmsvals = []
+        confIds = [conf.GetId() for conf in mol.GetConformers()]
+        if not prealigned:
+            if atom_id_in_ring:
+                AlignMolConformers(mol, atomIds=atom_id_in_ring, RMSlist=rmsvals)
+            else:
+                AlignMolConformers(mol, RMSlist=rmsvals)
+        else:  # already prealigned
+            for i in range(1, len(confIds)):
+                rmsvals.append(GetConformerRMSFromAtomIds(mol, confIds[0], confIds[i], atomIds=atom_id_in_ring, prealigned=prealigned))
+        # loop over the conformations (except the reference one)
+        cmat_per_ring = []
+        for i in range(1, len(confIds)):
+            cmat_per_ring.append(rmsvals[i - 1])
+            for j in range(1, i):
+                cmat_per_ring.append(GetConformerRMSFromAtomIds(mol, confIds[i], confIds[j], atomIds=atom_id_in_ring, prealigned=True))
+
+        cmat_list.append(np.array(cmat_per_ring))
+
+    cmat_list_array = np.array(cmat_list)
+
+    return list(np.mean(cmat_list_array, axis=0))
```

Comment:
I'm not strong in numpy to write the code directly from the top of my head. So, I'll explain. If `cmat_list_array` is a matrix A of size N x M, where N is a number of conformer pairs and M is the number of rings. You have to compute A^2 for every element, then multiply each column on the corresponding number of atoms in ring 1..M, sum over rows, divide on the sum of all ring atoms in a molecule and take a square root. This would be mathematically exact.
Simple averaging will work only if all rings have the same size.

Reply:
I am still confused, but do you want to calculate the mathematical RMSD based on the total number of atoms in the ring? Say we have a five- and six- membered ring. Instead of calculating:
$$\frac{\sqrt{\frac{\sum{d_{a}^{2}}}{5}} + \sqrt{\frac{\sum{d_{b}^{2}}}{6}}}{2} $$
We should calculate it this way to reflect the actual RMSD:
$$\sqrt{\frac{\sum{d_{a}^{2}} + \sum{d_{b}^{2}}}{5 + 6}}$$

Reply:
Yes, I propose to calculate true RMSD which will account atoms in all ring systems - the latter equation. However, we have RMSD for individual rings only.

$$RMSD_{a} = \sqrt{\frac{\sum{d_{a}^{2}}}{5}}$$

and 

$$RMSD_{b} = \sqrt{\frac{\sum{d_{b}^{2}}}{5}}$$

Therefore, I suggest to recalculate them into the final RMSD by simple math operations.

Reply:
alright. I have edited the code to calculate the true RMSD based on the rdkit RMSD